Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,477 @@
+    // being uploaded.
+    //
+    // TODO: We need to finalize which formats we support.
+    private class func inputImageUTISet() -> Set<String>! {
+        return [
+            kUTTypeJPEG as String,
+            kUTTypeGIF as String,
+            kUTTypePNG as String
+        ]
+    }
+
+    // Returns the set of UTIs that correspond to valid _output_ image formats
+    // for Signal attachments.
+    //
+    // TODO: We need to finalize which formats we support.
+    private class func outputImageUTISet() -> Set<String>! {
+        return [
+            kUTTypeJPEG as String,
+            kUTTypeGIF as String,
+            kUTTypePNG as String
+        ]
+    }
+
+    // Returns the set of UTIs that correspond to valid video formats
+    // for Signal attachments.
+    //
+    // TODO: We need to finalize which formats we support.
+    private class func videoUTISet() -> Set<String>! {
+        return [
+            kUTTypeMPEG4 as String
+        ]
+    }
+
+    // Returns the set of UTIs that correspond to valid audio formats
+    // for Signal attachments.
+    //
+    // TODO: We need to finalize which formats we support.
+    private class func audioUTISet() -> Set<String>! {
+        return [
+            kUTTypeMP3 as String,
+            kUTTypeMPEG4Audio as String
+        ]
+    }
+
+    // Returns the set of UTIs that correspond to valid input formats
+    // for Signal attachments.
+    public class func validInputUTISet() -> Set<String>! {
+        return inputImageUTISet().union(videoUTISet().union(audioUTISet()))
+    }
+
+    public func isImage() -> Bool {
+        return SignalAttachment.outputImageUTISet().contains(dataUTI)
+    }
+
+    public func isVideo() -> Bool {
+        return SignalAttachment.videoUTISet().contains(dataUTI)
+    }
+
+    public func isAudio() -> Bool {
+        return SignalAttachment.audioUTISet().contains(dataUTI)
+    }
+
+    // Returns an attachment from the pasteboard, or nil if no attachment
+    // can be found.
+    //
+    // NOTE: The attachment returned by this method may not be valid.
+    //       Check the attachment's error property.
+    public class func attachmentFromPasteboard() -> SignalAttachment? {
+        guard UIPasteboard.general.numberOfItems == 1 else {
+            // Ignore pasteboard if it contains multiple items.
+            //
+            // TODO: Should we try to use the first?
+            return nil
+        }
+        let pasteboardUTISet = Set(UIPasteboard.general.types)
+        for dataUTI in inputImageUTISet() {
+            if pasteboardUTISet.contains(dataUTI) {
+                let imageData = UIPasteboard.general.data(forPasteboardType:dataUTI)
+                return imageAttachment(withData : imageData, dataUTI : dataUTI)
+            }
+        }
+        for dataUTI in videoUTISet() {
+            if pasteboardUTISet.contains(dataUTI) {
+                let imageData = UIPasteboard.general.data(forPasteboardType:dataUTI)
```

Comment:
> nit: This should probably be let videoData = ... same with let audioData below.

Just going to use `data`.  

> Or maybe there's some way to dedupe this...

We need to find the "best" dataUTI and its corresponding data.  Deduplicating that logic would mean an optional tuple, and unpacking an optional tuple would be more verbose, unless I'm missing something.